### PR TITLE
Add related ops & txs; 20 chain remediations

### PR DESCRIPTION
**Pull request by LindaOrtega**
---

- [x] **Add Rosetta metadata types**: This makes it easier to later document and expand the different metadatas used.  
Specifically, added `OperationMetaData`, `AccountIdMetaData`, `ContinuationMetaData`, `TransactionMetaData`.

---

- [x] **More transparency on related transactions**: The only related transactions in Chainweb are continuations (both in the same chain and cross-chain). 
Therefore, the `ContinuationMetaData` and `TransactionMetaData` provide more transparency into a continuation's current step and the chain id of the next step (if there is a next step). It also provides the request key of the transaction that originated the continuation (i.e. the pact id). 

(Some limitations): 
- It doesn't provide information on previous steps
- the chain id where the continuation was initiated (unless its the first step)
- if the current step triggered a rollback

_Motivation_: https://www.rosetta-api.org/docs/models/Transaction.html

---

- [x] **More transparencies on related operations**: Operations within a transaction are now linked using the `Operation` type's `related_operations` field to represent a DAG of operations. Gas operations are linked together, while coin-table operations are linked to each other in the order they appear.

_Motivation_: https://www.rosetta-api.org/docs/models/Operation.html

- [x] Refactor `UnindexedOperation` be a function that takes an operation index and a list of related operations to create an `Operation`. This is necessary because an operation index and related operations can only be determined once all operations in a transaction are known.
- [x] Add new `UnindexedOperations` type to keep track of which operations funded gas, which involved the user interacting with the coin-table, and which paid/redeemed gas. This facilitates linked fund and gas operations to each other.
- [x] Now testing operation index and related operations in Rosetta unit tests.

---

~~- [x] **Update to Rosetta 1.4.4**: This version includes changes to some data types and the addition of Construction API types. The latter API will be implemented in a follow up PR.~~
    ~~- [x] `RosettaError` now allows providing (optional) details on the error. For simplicity, we're not utilizing this feature. I also wanted to keep the enum functionality of the rosetta error type we've been using to this point.~~
    ~~- [x] Some endpoints' request and response schemas changed.~~

Most of these changes were already merged in https://github.com/kadena-io/chainweb-node/pull/1149. 

---

- [x] **Improved Remediation Log parsing/matching** (for both coin v2 and 20 chain). 
Using the TxId of the coinbase transaction and the list of remediation `Command`, the new algorithm can derive the TxId of these remediation txs with a couple of assumptions.

(Some assumptions):
1. All remediations are successful
2. All coinbase txs in remediation blocks are successful
3. Each remediation tx incremented the TxId counter by one. 

With this information, we can now match coin table logs to remediation txs (if they produced any) using a variation of the matching algorithm used for non-genesis transactions.

   - [x] Refactored Rosetta API tests in /block to reflect these changes.
   - [x] Adds 20 chain remediation in Chain 3 for version `FastTimedCPM` to allow simulating this remediation in the tests.
   - [x] Adds Rosetta API tests for checking coin v2 and 20 chain remediations in /block endpoint.
   - [x] Adds Rosetta API tests so that user transactions (i.e. transfers) make it into coin v2 and 20 chain remediation blocks. These edge cases would otherwise be hard to test.

---

   - [x] Adds balance delta, operation status, and account metadata testing to Rosetta API /block and /block/transaction tests.

---

   - [x] Adds metadata field for denoting which version of the implementation is being used. This is meant to capture if something changes with how rosetta is implemented internally. Needed to add this because the new remediation logic for coin v2 is not backwards compatible with previous implementations. Previously, the logs in the coin v2 block were all grouped together into a single request key, but now it will show the different request keys of the remediation transactions.

---

- [x] **Fix TxLog parsing error** caused by the fork to Pact 4.2.0 and adds test for it (see commit [here](https://github.com/kadena-io/chainweb-node/pull/1135/commits/15087fd892584936d1a6b8f3e1e4b8852ba6f669)).

---

- [x] Edit Changelog

### CODE REVIEW

File: src/Chainweb/Rosetta/Internal.hs [hunk 1]
```diff
@@ -351,66 +395,94 @@ nonGenesisTransaction logs initial rest target
...
+  where
+    makeAcc restLogs rosettaTx =
+      accumulatorFunction accTyp acc restLogs rosettaTx
+    makeOps (tid, logs) =
+      map (operation Remediation TransferOrCreateAcct tid) logs
+
+-- | Matches all transactions in a remediation block (including coinbase, remediations,
+--   and user transactions) to their coin table logs.
+--  Matches coinbase logs first, then remediations (assumes that each remediation transaction
+--  incremented the TxId counter), and uses same algorithm as non-genesis transaction matching
```
Comment (LindaOrtega) [inline src/Chainweb/Rosetta/Internal.hs:435]: Is this assumption correct? Under the hood, the remediations use the applyGenesis helper function from TransactionExec.

File: test/Chainweb/Test/Rosetta/RestAPI.hs [hunk 2]
```diff
@@ -209,49 +230,136 @@ blockTests tio envIo = testCaseSchSteps "Block Tests" $ \step -> do
...
+    _blockResp_otherTransactions resp @?= Nothing
+    Just b <- pure $ (_blockResp_block resp)
+    _block_metadata b @?= Nothing
+    _blockId_index (_block_blockId b) @?= bhChain20Rem
+    _blockId_index (_block_parentBlockId b) @?= (bhChain20Rem - 1)
+
+    case (_block_transactions b) of
+      x:y:_ -> do
+        step "check remediation transactions' request keys"
+        -- TODO: are these unique across lifetime of blockchain/chains?
```
Comment (LindaOrtega) [inline test/Chainweb/Test/Rosetta/RestAPI.hs:346]: @emilypi is this correct? Are the request keys for the remediations unique across the lifetime of the blockchain and across chains? Unique transaction id is required for rosetta and usually that's not a problem but remediations command payloads are not included into the block like coinbase/user transactions are. So I wanted to make sure if this assumption was correct?

Comment (emilypi) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:346]: I'm pretty sure that request keys for remediations are fixed across all chains, yes. Otherwise, we'd not be able to replay them. They do go into the hash of the block at that height iirc.

File: test/Chainweb/Test/Rosetta/RestAPI.hs [hunk 3]
```diff
@@ -209,49 +230,136 @@ blockTests tio envIo = testCaseSchSteps "Block Tests" $ \step -> do
...
+              [x,r1,y] -> do
+                -- ^ 20 chain remediation block
+                let ops = _transaction_operations x <> _transaction_operations r1 <>
+                          _transaction_operations y
+                case ops of
+                  [a,rop1, b',c,d,e,f] -> validateTxs (Just rop1) a b' c d e f
+                  _ -> assertFailure "should have 7 ops: coinbase + 20 chain rem + 5 for transfer tx"
+
+              [x,y] -> do
+                -- ^ not a remediation block
```
Comment (emilypi) [inline test/Chainweb/Test/Rosetta/RestAPI.hs:263]: -- ^ won't pass CI I don't think. It's haddock-specific syntax in the wrong position.

Comment (LindaOrtega) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:263]: What do you mean?

File: test/Chainweb/Test/Rosetta/RestAPI.hs (hunk 2 above)
Comment (LindaOrtega) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:346]: I thought only the logs of the remediations went into the hash?
I guess my question was is it possible that we could get a user-generated transaction that has the same request key as one of the remediations?

Comment (emilypi) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:346]: I'm not sure if it's only the logs, since we're calling the actual pact code and affecting the pact db environment. It may live exclusively in there, but I'd have to double check. As far as request keys... no. The request keys are generated by a hash of the transaction data and a nonce, and we track nonces to ensure they're unique - so request keys can be considered unique

File: test/Chainweb/Test/Rosetta/RestAPI.hs (hunk 3 above)
Comment (emilypi) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:263]: `-- ^` is special haddock syntax, which is reserved for top-level declarations. When we go to build docs, I believe that syntax will fail.

File: src/Chainweb/Rosetta/Internal.hs [hunk 4]
```diff
@@ -264,11 +301,15 @@ gasTransactionAcc accTyp txa@(TxAccumulator logs' _) ctx = combine logs'
...
     combine [] = listErr "No logs found" ([] :: [(TxId, [AccountLog])])
 
-    makeAcc restLogs fund transfer gas = pure $
-      accumulatorFunction accTyp txa restLogs tx
+    makeAcc restLogs fund transfer gas = pure $!
+      accumulatorFunction accTyp acc restLogs tx
       where
-        tx = makeRosettaTx ctx $ indexedOperations $
-             fund <> transfer <> gas
+        tx = makeRosettaTx ctx cid $! indexedOperations $!
```
Comment (LindaOrtega) [inline src/Chainweb/Rosetta/Internal.hs:307]: All the changes that add `ChainId` is because its consumed here by `makeRosettaTx` to give more insight into related transactions.

File: test/Chainweb/Test/Rosetta/RestAPI.hs (hunk 3 above)
Comment (LindaOrtega) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:263]: aw, I've used `-- ^`  in other places as well (mainly in documenting specific type's fields). Will it break in those cases as well?

Comment (emilypi) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:263]: possibly. if you build the docs you can see.

File: src/Chainweb/Rosetta/Internal.hs [hunk 5]
```diff
@@ -351,66 +383,105 @@ nonGenesisTransaction logs initial rest target
...
+    -- ^ NOTE: No CommandResult available for these.
+    -> V.Vector (CommandResult Hash)
+    -- ^ User transactions in the same block as remediations
+    -> Either String [Transaction]
+remediations logs cid coinbase remTxs txs = do
+  TxAccumulator restLogs coinbaseTx <- nonGenesisCoinbaseLog logsList cid coinbase
+  coinbaseTxId <- note "remediations: No TxId found for Coinbase" (_crTxId coinbase)
+
+  let remWithTxIds = zip remTxs [(succ coinbaseTxId)..]
+      -- ^ Assumes that each remediation transaction gets its own TxId
```
Comment (emilypi) [inline src/Chainweb/Rosetta/Internal.hs:451]: And assumes that txids are going to be sequential for each remediation. Is this a correct assumption?

File: src/Chainweb/Rosetta/Utils.hs [hunk 6]
```diff
@@ -42,6 +45,151 @@ import Chainweb.Version
...
+  -- TODO: Add check to see if curr step was rolled back.
+  --       This would also mean a next step is not occuring.
+  | otherwise = case P._peYield pe >>= P._yProvenance of
+      Nothing -> Just $ ContinuationNextStep $ chainIdToText currChainId
+      -- ^ next step occurs in the same chain
+      Just (P.Provenance nextChainId _) ->
+      -- ^ next step is a cross-chain step
+        Just $ ContinuationNextStep (P._chainId nextChainId)
+  where
+    isLastStep = succ $ P._peStep pe == P._peStepCount pe
```
Comment (emilypi) [inline src/Chainweb/Rosetta/Utils.hs:191]: Does this need to be 

```haskell
isLastStep = (succ P._peStep pe) == P._peStepCount pe
```
?

Or does `==` bind more loosely than `succ`? Either way, should probably explicitly paren this.

File: test/Chainweb/Test/Rosetta/RestAPI.hs [hunk 7]
```diff
@@ -76,6 +78,20 @@ cids = chainIds v ^.. folded . to chainIdInt . to (sshow @Int)
 
 nonceRef :: IORef Natural
 nonceRef = unsafePerformIO $ newIORef 0
+{-# NOINLINE nonceRef #-}
```
Comment (emilypi) [inline test/Chainweb/Test/Rosetta/RestAPI.hs:81]: :+1:

**Review by emilypi — COMMENTED** [review]
This looks good to me. Have we approved using the Rosetta test harness they supplied us?

File: src/Chainweb/Rosetta/Utils.hs (hunk 6 above)
Comment (emilypi) [reply src/Chainweb/Rosetta/Utils.hs:191]: Followup: 

```haskell
П> succ False
True
П> succ True
*** Exception: Prelude.Enum.Bool.succ: bad argument
П> succ $ True == False
True
П> succ $ True == True
*** Exception: Prelude.Enum.Bool.succ: bad argument
```

Yeah, that looks like a potential bomb waiting to go off to me.

File: test/Chainweb/Test/Rosetta/RestAPI.hs (hunk 3 above)
Comment (LindaOrtega) [reply test/Chainweb/Test/Rosetta/RestAPI.hs:289]: Do you know how to build the docs? I tried `cabal haddock` and it just built the docs for the libraries

File: src/Chainweb/Rosetta/Utils.hs (hunk 6 above)
Comment (LindaOrtega) [reply src/Chainweb/Rosetta/Utils.hs]: You're right. I made it more explicit via:
```
nextStep = succ $ P._peStep pe
totalSteps = P._peStepCount pe
isLastStep = nextStep == totalSteps
```
I was getting a hlint warning when I added the explicit parems, but I think this solution makes it explicit and gets rid of that hlint warning.

**Comment by LindaOrtega** [comment]
@emilypi regarding this comment:
> Have we approved using the Rosetta test harness they supplied us?
Could you elaborate what you mean by this?

File: src/Chainweb/Rosetta/Internal.hs (hunk 5 above)
Comment (LindaOrtega) [reply src/Chainweb/Rosetta/Internal.hs]: Great question! I think I was also asking myself the same thing (see this comment https://github.com/kadena-io/chainweb-node/pull/1135#discussion_r482517990).

I also added this comment to the function description for `remediations`:
> NOTE: Assumes that each remediation transaction incremented the TxId counter by one.

I know that for the remediations that did occur this assumption seemed to hold.

This question might be a good candidate for pair reviewing session though.

Future remediations might not need to use this logic if we realize that Events give us all the information we need.

File: src/Chainweb/Rosetta/Internal.hs (hunk 1 above)
Comment (LindaOrtega) [reply src/Chainweb/Rosetta/Internal.hs]: Resolving this comment as it's a duplicate of the conversation occurring in https://github.com/kadena-io/chainweb-node/pull/1135#discussion_r798060937

File: src/Chainweb/Rosetta/Internal.hs [hunk 8]
```diff
@@ -118,73 +108,90 @@ matchLogs
     -> M.Map TxId [AccountLog]
     -> CoinbaseTx (CommandResult Hash)
     -> V.Vector (CommandResult Hash)
-    -> Either RosettaFailure tx
+    -> ExceptT RosettaFailure Handler tx
 matchLogs typ bh logs coinbase txs
   | bheight == genesisHeight v cid = matchGenesis
-  | coinV2Upgrade v cid bheight = matchRemediation
+  | coinV2Upgrade v cid bheight = matchRemediation upgradeTransactions
+  | to20ChainRebalance v cid bheight = matchRemediation twentyChainUpgradeTransactions
```
Comment (emilypi) [inline src/Chainweb/Rosetta/Internal.hs:115]: TODO: add coinv3 upgrade txs

File: src/Chainweb/Rosetta/Internal.hs [hunk 9]
```diff
@@ -351,66 +383,106 @@ nonGenesisTransaction logs initial rest target
...
+    -- User transactions in the same block as remediations
+    -> V.Vector (CommandResult Hash)
+    -> Either String [Transaction]
+remediations logs cid coinbase remTxs txs = do
+  TxAccumulator restLogs coinbaseTx <- nonGenesisCoinbaseLog logsList cid coinbase
+  coinbaseTxId <- note "remediations: No TxId found for Coinbase" (_crTxId coinbase)
+
+  let remWithTxIds = zip remTxs [(succ coinbaseTxId)..]
+      -- Assumes that each remediation transaction gets its own TxId.
+      -- Assumes that TxIds are going to be sequential for each remediation.
```
Comment (emilypi) [inline src/Chainweb/Rosetta/Internal.hs:452]: This assumption holds, since the remediation txs are applied _directly after coinbase at a particular height_, as part of `applyCoinbase`. We construct the blocks, hence, the txids are not random.

File: src/Chainweb/Rosetta/Internal.hs (hunk 5 above)
Comment (LindaOrtega) [reply src/Chainweb/Rosetta/Internal.hs]: Resolved via: 
> This assumption holds, since the remediation txs are applied directly after coinbase at a particular height, as part of applyCoinbase. We construct the blocks, hence, the txids are not random.

See https://github.com/kadena-io/chainweb-node/blob/c0c300a64040390d603f1183eac126e3bbfebe8d/src/Chainweb/Pact/TransactionExec.hs#L328

File: src/Chainweb/Rosetta/Internal.hs (hunk 8 above)
Comment (LindaOrtega) [reply src/Chainweb/Rosetta/Internal.hs]: Done, going to run rosetta-cli tests soon

Comment (LindaOrtega) [reply src/Chainweb/Rosetta/Internal.hs]: But did add local coinv3 block test

**Comment by LindaOrtega** [comment]
### **------ TEST COVERAGE ------**

Sample config file used: [configTestnetChain0.log](https://github.com/kadena-io/chainweb-node/files/8028497/configTestnetChain0.log)
Each chainId will have a different value for `network.sub_network_identifier.network`  and each version will have a different value for `network.network`.
Testnet is `testnet04` and mainnet is `mainnet01`.

#### **_Testnet, Chain 0 (from genesis to the current tip):_**
```
$ ./bin/rosetta-cli --configuration-file configs/testnet/configTestnetChain0.json check:data

...

Success: Reconciliation Coverage End Condition [Coverage: 100.000000%]

+--------------------+--------------------------------+--------+
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepencies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |   VALUE    |
+--------------------------+--------------------------------+------------+
| Blocks                   | # of blocks synced             |    1931822 |
+--------------------------+--------------------------------+------------+
| Orphans                  | # of blocks orphaned           |          0 |
+--------------------------+--------------------------------+------------+
| Transactions             | # of transaction processed     |    1947915 |
+--------------------------+--------------------------------+------------+
| Operations               | # of operations processed      |    1998980 |
+--------------------------+--------------------------------+------------+
| Accounts                 | # of accounts seen             |       7897 |
+--------------------------+--------------------------------+------------+
| Active Reconciliations   | # of reconciliations performed |    1810974 |
|                          | after seeing an account in a   |            |
|                          | block                          |            |
+--------------------------+--------------------------------+------------+
| Inactive Reconciliations | # of reconciliations performed |   11565815 |
|                          | on randomly selected accounts  |            |
+--------------------------+--------------------------------+------------+
| Exempt Reconciliations   | # of reconciliation failures   |          0 |
|                          | considered exempt              |            |
+--------------------------+--------------------------------+------------+
| Failed Reconciliations   | # of reconciliation failures   |          0 |
+--------------------------+--------------------------------+------------+
| Skipped Reconciliations  | # of reconciliations skipped   |     115740 |
+--------------------------+--------------------------------+------------+
| Reconciliation Coverage  | % of accounts that have been   | 99.848044% |
|                          | reconciled                     |            |
+--------------------------+--------------------------------+------------+
```


#### **_Testnet, Chain 0 (from 1260000 to 1262000):_**
Note that this test does not test the entire history. It runs the check during the range of block heights that the coin v3 fork occurs (block height 1,261,000 for testnet). This was necessary since the change of making the coin v3 its own transaction was added recently and the database that powered the more exhaustive test above has been running for a couple of weeks now. 

```
Success: Index End Condition [Index: 1262000]

+--------------------+--------------------------------+--------+
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepencies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+-------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |    VALUE    |
+--------------------------+--------------------------------+-------------+
| Blocks                   | # of blocks synced             |        2001 |
+--------------------------+--------------------------------+-------------+
| Orphans                  | # of blocks orphaned           |           0 |
+--------------------------+--------------------------------+-------------+
| Transactions             | # of transaction processed     |        2077 |
+--------------------------+--------------------------------+-------------+
| Operations               | # of operations processed      |        2230 |
+--------------------------+--------------------------------+-------------+
| Accounts                 | # of accounts seen             |          10 |
+--------------------------+--------------------------------+-------------+
| Active Reconciliations   | # of reconciliations performed |        2042 |
|                          | after seeing an account in a   |             |
|                          | block                          |             |
+--------------------------+--------------------------------+-------------+
| Inactive Reconciliations | # of reconciliations performed |          28 |
|                          | on randomly selected accounts  |             |
+--------------------------+--------------------------------+-------------+
| Exempt Reconciliations   | # of reconciliation failures   |           0 |
|                          | considered exempt              |             |
+--------------------------+--------------------------------+-------------+
| Failed Reconciliations   | # of reconciliation failures   |           0 |
+--------------------------+--------------------------------+-------------+
| Skipped Reconciliations  | # of reconciliations skipped   |           0 |
+--------------------------+--------------------------------+-------------+
| Reconciliation Coverage  | % of accounts that have been   | 100.000000% |
|                          | reconciled                     |             |
+--------------------------+--------------------------------+-------------+
```

#### **_Testnet, Chain 9 (from genesis to current tip):_**

```
$ ./bin/rosetta-cli --configuration-file configs/testnet/configTestnetChain9.json check:data

...

Success: Reconciliation Coverage End Condition [Coverage: 100.000000%]

+--------------------+--------------------------------+--------+
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepencies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |   VALUE    |
+--------------------------+--------------------------------+------------+
| Blocks                   | # of blocks synced             |    1931879 |
+--------------------------+--------------------------------+------------+
| Orphans                  | # of blocks orphaned           |          2 |
+--------------------------+--------------------------------+------------+
| Transactions             | # of transaction processed     |    1932486 |
+--------------------------+--------------------------------+------------+
| Operations               | # of operations processed      |    1939040 |
+--------------------------+--------------------------------+------------+
| Accounts                 | # of accounts seen             |       5187 |
+--------------------------+--------------------------------+------------+
| Active Reconciliations   | # of reconciliations performed |    1797353 |
|                          | after seeing an account in a   |            |
|                          | block                          |            |
+--------------------------+--------------------------------+------------+
| Inactive Reconciliations | # of reconciliations performed |    9252787 |
|                          | on randomly selected accounts  |            |
+--------------------------+--------------------------------+------------+
| Exempt Reconciliations   | # of reconciliation failures   |          0 |
|                          | considered exempt              |            |
+--------------------------+--------------------------------+------------+
| Failed Reconciliations   | # of reconciliation failures   |          0 |
+--------------------------+--------------------------------+------------+
| Skipped Reconciliations  | # of reconciliations skipped   |     109434 |
+--------------------------+--------------------------------+------------+
| Reconciliation Coverage  | % of accounts that have been   | 99.980721% |
|                          | reconciled                     |            |
+--------------------------+--------------------------------+------------+
```

**Review by emilypi — APPROVED** [review]
